### PR TITLE
trigger slot extraction in interactive mode

### LINKED ISSUE #10053: rasa forms in 3.x when slots are `from_text` differ behavior from 2.x
I'm in the process of rewriting my 2.x form examples into 3.x. While working on [this example](https://github.com/RasaHQ/rasa-2.x-form-examples/tree/main/03-simple-form) I think I may have stumbled upon a bug. 

The goal of the form is to retrieve a first and last name by fetching the slots from text. There are no entities being detected here. 

## Old Domain.yml

Originally, this was the form definition. 

```yml
forms:
  name_form:
    first_name:
    - type: from_text
    last_name:
    - type: from_text

slots:
  first_name:
    type: text
    influence_conversation: true
  last_name:
    type: text
    influence_conversation: true
```

This is the new definition, per 3.0 specs.

```yml
slots:
  first_name:
    type: text
    influence_conversation: true
    mappings:
      - type: from_text
  last_name:
    type: text
    influence_conversation: true
    mappings:
      - type: from_text

forms:
  name_form:
    required_slots:
      - first_name
      - last_name
```

Everything else is kept the same, however. I've made no changes to the stories/rule or nlu files.

## Differences

I'm trying out both approaches inside of `rasa interactive` but the behavior seems inconsistent. 

### Rasa 2.x Behavior

```
Chat History

 #    Bot                                                       You        
───────────────────────────────────────────────────────────────────────────
 1    action_listen                                                        
───────────────────────────────────────────────────────────────────────────
 2                                                                   hello 
                                                        intent: greet 1.00 
───────────────────────────────────────────────────────────────────────────
 3    utter_greet 1.00                                                     
      Hey! How are you?                                                    
      action_listen 1.00                                                   
───────────────────────────────────────────────────────────────────────────
 4                                                  i want to give my name 
                                                intent: request_names 1.00 
───────────────────────────────────────────────────────────────────────────
 5    name_form 1.00                                                       
      active_loop{"name": "name_form"}                                     
      slot{"requested_slot": "first_name"}                                 
      What is your first name?                                             
      slot{"first_name": "vincent"}                                        
      slot{"requested_slot": "last_name"}                                  


Current slots: 
        first_name: vincent, last_name: None, requested_slot: last_name, session_started_metadata: None
```

It's able to pick up the slots just fine. No "bothering" about intents. 

### Rasa 3.x Behavior

This is different in 3.x. It now doesn't pick up the slots anymore. 

```
 #    Bot                                                       You        
───────────────────────────────────────────────────────────────────────────
 1    action_listen                                                        
───────────────────────────────────────────────────────────────────────────
 2                                                                   hello 
                                                        intent: greet 1.00 
───────────────────────────────────────────────────────────────────────────
 3    utter_greet 1.00                                                     
      Hey! How are you?                                                    
      action_listen 1.00                                                   
───────────────────────────────────────────────────────────────────────────
 4                                                  i want to give my name 
                                                intent: request_names 0.99 
───────────────────────────────────────────────────────────────────────────
 5    name_form 1.00                                                       
      active_loop{"name": "name_form"}                                     
      slot{"requested_slot": "first_name"}                                 
      What is your first name?                                             


Current slots: 
        first_name: None, last_name: None, requested_slot: first_name, session_started_metadata: None

------
? The bot wants to run 'action_listen', correct? Yes                                                                                              
? Your input -> vincent                                                                                                                           
? Your NLU model classified 'vincent' with intent 'deny' and there are no entities, is this correct? (Y/n)
```

This gives me the following chat history, even though the input is exactly the same. 

```
Chat History

 #    Bot                                                       You        
───────────────────────────────────────────────────────────────────────────
 1    action_listen                                                        
───────────────────────────────────────────────────────────────────────────
 2                                                                   hello 
                                                        intent: greet 1.00 
───────────────────────────────────────────────────────────────────────────
 3    utter_greet 1.00                                                     
      Hey! How are you?                                                    
      action_listen 1.00                                                   
───────────────────────────────────────────────────────────────────────────
 4                                                  i want to give my name 
                                                intent: request_names 0.99 
───────────────────────────────────────────────────────────────────────────
 5    name_form 1.00                                                       
      active_loop{"name": "name_form"}                                     
      slot{"requested_slot": "first_name"}                                 
      What is your first name?                                             


Current slots: 
        first_name: None, last_name: None, requested_slot: first_name, session_started_metadata: None
```

Note: the first name slot is now no longer set. 

This feels like an inconsistency, it may be related to the [stories](https://github.com/RasaHQ/rasa-2.x-form-examples/blob/main/03-simple-form/data/stories.yml) but if that's the case then we may need to be more explicit about that in migration documents.

### LINKED ISSUE #10213: `rasa interactive` stuck in infinite loop in Rasa 3.0rc1
I'm trying to teach my assistant to handle a form correctly. In particular, I have a form that has these slots: 

```yaml
forms:
  fancy_pizza_form:
    required_slots:
    - vegetarian
    - pizza_size
    - pizza_type
slots:
  pizza_size:
    type: text
    influence_conversation: true
    mappings:
    - type: from_entity
      entity: pizza_size
  pizza_type:
    type: text
    influence_conversation: true
    mappings:
    - type: from_entity
      entity: pizza_type
  vegetarian:
    type: bool
    influence_conversation: true
    mappings:
    - type: from_intent
      value: true
      intent: affirm
    - type: from_intent
      value: false
      intent: deny
```

I don't have stories for this form yet, so I figured I'd add one via `rasa interactive`. I'm stuck at this step. 

```
------
Chat History

 #    Bot                                                                You        
────────────────────────────────────────────────────────────────────────────────────
 1    action_listen                                                                 
────────────────────────────────────────────────────────────────────────────────────
 2                                                                         hi there 
                                                                 intent: greet 1.00 
────────────────────────────────────────────────────────────────────────────────────
 3    utter_greet 1.00                                                              
      Hey! How are you?                                                             
      action_listen 1.00                                                            
────────────────────────────────────────────────────────────────────────────────────
 4                                                      i want to buy a fancy pizza 
                                                       intent: buy_fancy_pizza 1.00 
────────────────────────────────────────────────────────────────────────────────────
 5    fancy_pizza_form 1.00                                                         
      active_loop{"name": "fancy_pizza_form"}                                       
      slot{"requested_slot": "vegetarian"}                                          
      Would you like to order a vegetarian pizza?                                   
      Buttons:                                                                      
      1: yes (/affirm)                                                              
      2: no (/deny)                                                                 
      Type out your own                                                             
      message...                                                                    
      action_listen 1.00                                                            
────────────────────────────────────────────────────────────────────────────────────
 6                                                                            /deny 
                                                                  intent: deny 1.00 


Current slots: 
        pizza_size: None, pizza_type: None, vegetarian: None, requested_slot: vegetarian, session_started_metadata: None

------
? The bot wants to run 'action_default_fallback', correct? No                                                                                      
------
```

There's a fallback triggered! That's no bueno. So I'll try to correct it. 

```
? What is the next action of the bot? (Use arrow keys)                                                                                             
   <create new action>
   0.30 action_default_fallback                 
   0.00 ...                                     
   0.00 action_ask_pizza_type                   
   0.00 action_ask_vegetarian                   
   0.00 action_back                             
   0.00 action_deactivate_loop                  
   0.00 action_default_ask_affirmation 
...
 » 0.00 fancy_pizza_form                        
```

This brings up a prompt. So far so good. 

```
? What is the next action of the bot? 0.00 fancy_pizza_form                                                                                        
Thanks! The bot will now run fancy_pizza_form.

? Should 'fancy_pizza_form' validate user input to fill the slot 'vegetarian'? (Y/n)
```

I hit `Y` and I see ... 

```
Chat History

 #    Bot                                                                You        
────────────────────────────────────────────────────────────────────────────────────
 1    action_listen                                                                 
────────────────────────────────────────────────────────────────────────────────────
 2                                                                         hi there 
                                                                 intent: greet 1.00 
────────────────────────────────────────────────────────────────────────────────────
 3    utter_greet 1.00                                                              
      Hey! How are you?                                                             
      action_listen 1.00                                                            
────────────────────────────────────────────────────────────────────────────────────
 4                                                      i want to buy a fancy pizza 
                                                       intent: buy_fancy_pizza 1.00 
────────────────────────────────────────────────────────────────────────────────────
 5    fancy_pizza_form 1.00                                                         
      active_loop{"name": "fancy_pizza_form"}                                       
      slot{"requested_slot": "vegetarian"}                                          
      Would you like to order a vegetarian pizza?                                   
      Buttons:                                                                      
      1: yes (/affirm)                                                              
      2: no (/deny)                                                                 
      Type out your own                                                             
      message...                                                                    
      action_listen 1.00                                                            
────────────────────────────────────────────────────────────────────────────────────
 6                                                                            /deny 
                                                                  intent: deny 1.00 


Current slots: 
        pizza_size: None, pizza_type: None, vegetarian: None, requested_slot: vegetarian, session_started_metadata: None

------
? The bot wants to run 'action_default_fallback', correct? (Y/n)
```

Crud ... stuck in a loop. It feels like there's a bug around here. 

```
> python -m rasa --version
Rasa Version      :         3.0.0rc1
Minimum Compatible Version: 2.8.9
Rasa SDK Version  :         3.0.0rc1
Rasa X Version    :         None
Python Version    :         3.7.9
Operating System  :         Linux-5.11.0-7614-generic-x86_64-with-Pop-20.10-groovy
Python Path       :         /home/vincent/Development/how-to-rasa/venv/bin/python
```

### LINKED ISSUE #10220: `rasa interactive` does not apply slot mappings (3.0.0rc2)
### Rasa Open Source version

3.0.0rc2

### What happened?

Trained a moodbot with the addition of a `name` slot filled by an entity and a `yes` slot filled `from_intent`  i.e. these changes in domain.yml:
```
entities:
- name
slots:
  name:
    type: any
    influence_conversation: false
    mappings:
    - type: from_entity
      entity: name
  yes:
    type: any
    influence_conversation: false
    mappings:
    - type: from_intent
      intent: affirm
      value: yes
    - type: from_intent
      not_intent: affirm
      value: no
responses:
  utter_greet:
  - text: Hey {name}! How are you?
```
And some examples of `name` in nlu.yml:
```
nlu:
- intent: greet
  examples: |
    - hey [Molly](name)
    - hello [Bob](name)
    - hi [Chris](name)
    - hello there [Jim](name)
    - good morning [Fred](name)
    - good evening [Sally](name)
```

However neither the `name` nor the `yes` slot is filled during interactive learning.
The equivalent set up in 2.8.12 does work (`name` slot filled during interactive learning); the slots are also filled succesfully with `rasa shell` in 3.0.0rc2 - it's just `rasa interactive` that doesn't do it. 

### Command / Request

```shell
rasa interactive
```

Then 
```
hi Molly
```

It should end up with the same slot values as it does during `rasa shell`, but it ends up with nothing filled
```
2021-11-15 21:45:47 DEBUG    rasa.core.processor  - Current slot values: 
        name: Molly
        yes: no
        session_started_metadata: None
```


### Relevant log output

```shell
2021-11-15 21:29:13 INFO     rasa.core.training.interactive  - Restarted conversation, starting a new one.
? Your input -> hi Molly                                                                                                                                                                                                                                     
? Is the intent 'greet' correct for 'hi [Molly](name)' and are all entities labeled correctly? Yes                                                                                                                                                           
------
Chat History

 #    Bot                        You        
────────────────────────────────────────────
 1    action_listen                         
────────────────────────────────────────────
 2                         hi [Molly](name) 
                         intent: greet 1.00 


Current slots: 
        name: None, session_started_metadata: None
------
? The bot wants to run 'utter_greet', correct? Yes                                                                                                                                                                                                           
------
Chat History

 #    Bot                                 You        
─────────────────────────────────────────────────────
 1    action_listen                                  
─────────────────────────────────────────────────────
 2                                  hi [Molly](name) 
                                  intent: greet 1.00 
─────────────────────────────────────────────────────
 3    utter_greet 0.84                               
      Hey None! How are you?                         


Current slots: 
        name: None, session_started_metadata: None

------
? The bot wants to run 'action_listen', correct? Yes   
```
```

### PR DESCRIPTION
**Pull request by ka-bu**
**Proposed changes**:
- closes #10053 
- closes #10213
- closes #10220

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/server.py [hunk 1]
```diff
@@ -998,7 +998,14 @@ async def add_message(request: Request, conversation_id: Text) -> HTTPResponse:
 
         try:
             async with app.agent.lock_store.lock(conversation_id):
-                tracker = await app.agent.log_message(user_message)
+                # cf. processor.handle_message (ignoring prediction loop run)
+                tracker = await app.agent.processor.log_message(
+                    user_message, should_save_tracker=False
+                )
+                tracker = await app.agent.processor._run_action_extract_slots(
```
Comment (ancalita) [inline rasa/server.py:1005]: Would it be a good idea to make this method public with a simple docstring? Originally I made it private thinking it won't be used outside of the `MessageProcessor`, but here we are 😄 
Same consideration for `_save_tracker` too?

**Review by ancalita — COMMENTED** [review]
Amazing 🥇
To double-check do we have a test to make sure the tracker is updated in `test_server`?

File: rasa/server.py (hunk 1 above)
Comment (ka-bu) [reply rasa/server.py:1005]: oh good point, yes

Comment (ancalita) [reply rasa/server.py:1005]: Yeah might mean changing the method usage to public in several places in `MessageProcessor` 😄

**Comment by ka-bu** [comment]
> To double-check do we have a test to make sure the tracker is updated in test_server?

Oh, thanks! I thought I had checked the tests (whatever I had looked at didn't contain many tests and I thought it was out of scope to add new tests for everything) but there is a lot 🙈  

Looks like there is no test for "/messages" / `add_messages` yet, will add some in a bit 👍

File: rasa/core/processor.py [hunk 2]
```diff
@@ -125,25 +125,25 @@ async def handle_message(
...
 
         if isinstance(message.output_channel, CollectingOutputChannel):
             return message.output_channel.messages
 
         return None
 
-    async def _run_action_extract_slots(
+    async def run_action_extract_slots(
         self, output_channel: OutputChannel, tracker: DialogueStateTracker,
     ) -> DialogueStateTracker:
```
Comment (ancalita) [inline rasa/core/processor.py:148]: I'm going to predict code quality will fail because of lack of docstrings, both here and in `save_tracker`.

**Review by ancalita — APPROVED** [review]
Just one comment on docstrings to address otherwise good to go 🎉

File: rasa/core/processor.py (hunk 2 above)
Comment (ka-bu) [reply rasa/core/processor.py:148]: ouch, yeah, should've seen that, thanks :D

**Pull request merged**